lib/Sema/ConstraintSystem.cpp

```diff
@@ -2417,14 +2404,22 @@ bool ConstraintSystem::isConversionNonEphemeral(
         continue;
       }
 
+      // Look through load exprs. These can occur when projecting through a
+      // property on a class, which we want to support if the class is
+      // non-final.
```

Comment:
I don't think this is the correct condition. What case is this trying to handle?

Reply:
This is handling the following (which we want to allow):

```swift
final class C {
  var str = ""
}

struct S {
  var c = C()
}

var s = S()
let ptr = UnsafeMutablePointer(&s.c.str)
```

The base `s` is an lvalue, but we want an rvalue `C`, so a `LoadExpr` is inserted before then performing the member access for `str`.

Reply:
I don't think we want to allow this. Basically if you load from a class, you don't care what the base was anymore, and then either we allow forming pointers to instance variables or we don't. Being a global doesn't matter at that point. Right, @atrick?

Reply:
`let ptr = UnsafeMutablePointer(&s.c.str)`

If `s` is a `var`, then there's there's nothing keeping `c` alive. So escaping the addres of `c.str` seems pretty bad.

If `s` is a top-level or global `let` then `c` stays alive. However, technically we don't guarantee that a Swift class object has a persistent address (it may need to for satisfying some runtime constraint, but the language doesn't guarantee it). So I don't think we should allow that either.

Sorry if I missed that in my review.

Reply:
Looking through `diag_non_ephemeral_warning.swift` again, I see these two cases that still don't produce warnings, but probably should:
```
takesRaw(&topLevelFinalC.storedProperty)
takesRaw(&topLevelFragileFinalC.storedProperty)
```

Reply:
I think we probably do want to make that guarantee, but even without that differentiating between global `let` and global `var` seems way too subtle.

Reply:
We do support `realloc` on uniquely reference objects.

Reply:
@jrose-apple As I understand it, the base still matters for class instances. For example, we don't want to allow users to form a pointer to an instance variable on a class-typed local variable, as we don't guarantee it will remain initialised until the end of the stack frame.

Reply:
> `let ptr = UnsafeMutablePointer(&s.c.str)`
>
> If `s` is a `var`, then there's there's nothing keeping `c` alive. So escaping the addres of `c.str` seems pretty bad.

@atrick Should `c` not be guaranteed to be kept alive as long as `s.c` isn't changed to a different reference? IMO this would be reasonable constraint to place on the lifetime of the pointer created from the inout-to-pointer conversion.

> If `s` is a top-level or global `let` then `c` stays alive. However, technically we don't guarantee that a Swift class object has a persistent address (it may need to for satisfying some runtime constraint, but the language doesn't guarantee it). So I don't think we should allow that either.

Ah, okay then. I think for now at least I'll revert b6e1e996bd3c99e70c7c344d4c2ea5010cea4eb9, with the view of possibly reinstating it if we ever officially make such a guarantee. I somewhat doubt many (if any) people are using this pattern anyway – there was no sign of it in the source compat suite.

Reply:
> @atrick Should c not be guaranteed to be kept alive as long as s.c isn't changed to a different reference? IMO this would be reasonable constraint to place on the lifetime of the pointer created from the inout-to-pointer conversion.

Yes. If `s` is a `let`, then we have that guarantee. And on second thought, I don't think it will ever be possible to reallocate an object that has a `let` reference. But I think @jrose-apple and I agree that it's better to have a simple rule that disallows non-ephemeral pointers to instance variables.

Reply:
> Ah, okay then. I think for now at least I'll revert b6e1e99,

Wait a sec, that commit also allows non-ephemeral pointers to `static` variables. I think that's ok under the circumstances that you've identified.

Reply:
> Does this include non-ephemeral pointers to stored instance properties on global/static variable structures?

No, stored properties on structs can be pointed to if the struct itself is global/static. We want to disallow pointing to stored properties inside class instances, which I think we were calling "instance variables".